dbt/macros/dist_to_nearest_geometry.sql

```diff
@@ -0,0 +1,88 @@
+        -- exist for 2020, and distinct_years goes up to 2023, then park data
+        -- from 2020 will be filled forward to 2023
+        distinct_years as (select distinct year from {{ source("spatial", "parcel") }}),
+
+        -- Crosswalk of the source data and distinct years, used to perform
+        -- the forward filling described above
+        fill_years as (
+            select dy.year as pin_year, max(df.year) as fill_year
+            from {{ source_model }} as df
+            cross join distinct_years as dy
+            where
+                dy.year >= df.year
+                {% if source_conditional is defined %}
+                    and {{ source_conditional }}
+                {% endif %}
+            group by dy.year
+        ),
+
+        -- Source table with forward filling applied by year. This
+        -- will result in one row per geometry per year of parcel data i.e.
+        -- for hospitals, there will be one row for each hospital for each
+        -- year of parcel data (up to the current year)
+        location as (
+            select fy.pin_year, fill_data.*
+            from fill_years as fy
+            inner join {{ source_model }} as fill_data on fy.fill_year = fill_data.year
+            {% if source_conditional is defined %}
+                where {{ source_conditional }}
+            {% endif %}
+        ),
+
+        -- Source table with forward filling applied by year, but containing
+        -- the union of all geometries for each year. This will result in just
+        -- one record per year, with all geometries for each year squished into
+        -- a single object
+        location_agg as (
+            select
+                loc.pin_year,
+                max(loc.year) as fill_year,
+                geometry_union_agg(st_geomfrombinary(loc.geometry_3435)) as geom_3435
+            from location as loc
+            group by loc.pin_year
+        ),
+
+        -- For each unique PIN location, find the nearest point from each
+        -- geometry set from each year. The output of geometry_nearest_points
+        -- is a pair of points, one from each geometry and year
+        nearest_point as (
+            select
+                dp.x_3435,
+                dp.y_3435,
+                loc_agg.pin_year,
+                geometry_nearest_points(
+                    st_point(dp.x_3435, dp.y_3435), loc_agg.geom_3435
+                ) as points
+            from distinct_pins as dp
+            cross join location_agg as loc_agg
+        )
+
+    -- Using the nearest point from each target geometry and year, join the
+    -- nearest location data (name, id, etc.) to each PIN. Also calculate
+    -- distance between the nearest points
+    select
+        np.x_3435, np.y_3435, loc.*, st_distance(np.points[1], np.points[2]) as dist_ft
+    from nearest_point as np
+    inner join
+        location as loc
+        on st_intersects(np.points[2], st_geomfrombinary(loc.geometry_3435))
```

Comment:
It happens a lot! For example basically every CTA stop has two stop IDs for each unique point (one stop ID for each side of the street).

That's the reason I'm using `ARBITRARY` in the outer queries, it will grab one of the duplicate PIN/ID combinations. It's also the reason there are so many row-level differences: this query produces a different set of values to feed to `ARBITRARY`. I _think_ `ARBITRARY` is deterministic, but it's hard to confirm.

Reply:
Ah right, I forgot about the `ARBITRARY` clauses in the calling queries! From an interface perspective, it might be cleaner to do the deduplication inside the context of the `dist_to_nearest_geometry` macro, since as a macro consumer I would expect a "nearest" computation to return only one result. But that may be difficult given the way the queries are designed, and since there's a small universe of callers of this macro, I don't think it's a big deal to move forward with the understanding that it's possible for this macro to return multiple results for one PIN/year combo as long as it's called out in the docstring.

Reply:
Unfortunately, the columns within the `ARBITRARY()` clause are unique for each data set, and I can't think of a way to pass them to the outer query (no `ARBITRARY(xy.*) AS ...` exists). I agree it's a little untidy though :/